base/logging.jl

```diff
@@ -0,0 +1,509 @@
+module Logging
+
+import Base.Meta: isexpr
+
+export
+    # Frontend
+    @debug, @info, @warn, @error, @logmsg,
+    # Log control
+    with_logger, current_logger, global_logger,
+    disable_logging, configure_logging,
+    # Logger type
+    AbstractLogger,
+    # Logger methods
+    logmsg, shouldlog
+
+"""
+A logger controls how log records are filtered and dispatched.  When a log
+record is generated, the logger is the first piece of user configurable code
+which gets to inspect the record and decide what to do with it.
+"""
+abstract type AbstractLogger ; end
+
+
+"""
+Severity/verbosity of a log record.
+
+The log level provides a key against which potential log records may be
+filtered, before any other work is done to construct the log record data
+structure itself.
+"""
+struct LogLevel
+    level::Int32
+end
+
+LogLevel(level::LogLevel) = level
+
+Base.isless(a::LogLevel, b::LogLevel) = isless(a.level, b.level)
+Base.:+(level::LogLevel, inc) = LogLevel(level.level+inc)
+Base.:-(level::LogLevel, inc) = LogLevel(level.level-inc)
+
+const BelowMinLevel = LogLevel(typemin(Int32))
+const Debug         = LogLevel(-1000)
+const Info          = LogLevel(0)
+const Warn          = LogLevel(1000)
+const Error         = LogLevel(2000)
+const NoLogs        = LogLevel(typemax(Int32))
+
+function Base.show(io::IO, level::LogLevel)
+    if     level == BelowMinLevel  print(io, "BelowMinLevel")
+    elseif level == Debug          print(io, "Debug")
+    elseif level == Info           print(io, "Info")
+    elseif level == Warn           print(io, "Warn")
+    elseif level == Error          print(io, "Error")
+    elseif level == NoLogs         print(io, "NoLogs")
+    else                           print(io, "LogLevel($(level.level))")
+    end
+end
+
+parse_level(level) = level
+parse_level(level::String) = parse_level(Symbol(lowercase(level)))
+function parse_level(level::Symbol)
+    if      level == :belowminlevel  return  BelowMinLevel
+    elseif  level == :debug          return  Debug
+    elseif  level == :info           return  Info
+    elseif  level == :warn           return  Warn
+    elseif  level == :error          return  Error
+    elseif  level == :nologs         return  NoLogs
+    else
+        throw(ArgumentError("Unknown log level $level"))
+    end
+end
+
+
+# Global log limiting mechanism for super fast but inflexible global log
+# limiting.
+const _min_enabled_level = Ref(Debug)
+
+
+# A concretely typed cache of data extracted from the logger, plus the logger
```

Comment:
Most of these comments look like they should be docstrings.

Reply:
The convention I'm using is that implementation details have comments but no docstrings. How is this generally handled in `Base`? I don't want to give people the idea that `Logging.LogState` (for example) is part of the public API.

Reply:
That seems sensible. A quick browse shows me that there is a non-exported function in Enums.jl with a comment rather than a doc-string, so it seems similar to what you're doing.